File: plugins/prompt-creator/scripts/convert_format.js
#!/usr/bin/env node
// convert_format.js — 7層正規形YAML → md/json/xml/yaml 変換 (Node 標準のみ)
// 設計方針: 日本語7層・ゴールシーク構造を「構造保存」で変換する。
//   YAML 本文（key: value / リスト / - [ ] チェックリスト / 達成ゴール 等）を捨てず保持し、
//   Layer 見出しのみ提示形式に合わせて書き換える。これにより可逆で、ゴールシーク要素を落とさない。
// マーカーは scaffold_prompt.js / merge_layers.js が出力する「# Layer N: <title>」に一致させる。
"use strict";
const fs = require("fs");
const path = require("path");

function parseArgs(argv) {
  const args = { input: null, format: "md", output: null };
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === "--input") args.input = argv[++i];
    else if (argv[i] === "--format") args.format = argv[++i];
    else if (argv[i] === "--output") args.output = argv[++i];
  }
  return args;
}

// 「# Layer N: <title>」マーカーで本文を 7 層に分割し、本文を verbatim 保持する。
function parseLayers(text) {
  const layers = [];
  for (let n = 1; n <= 7; n++) {
    const re = new RegExp(
      `#+\\s*Layer\\s*${n}\\s*[:：]\\s*([^\\n]*)\\n([\\s\\S]*?)(?=#+\\s*Layer\\s*${n + 1}\\s*[:：]|$)`
    );
    const m = text.match(re);
    if (m) {
      layers.push({ n, title: m[1].trim(), body: m[2].replace(/\s+$/g, "") });
    }
  }
  return layers;
}

function toMd(layers) {
  const out = ["# 7層構造プロンプト", ""];
  for (const { n, title, body } of layers) {
    out.push(`## Layer ${n}: ${title}`, "");
    if (body.trim()) out.push(body.trimEnd(), "");
  }
  return out.join("\n");
}

function toJson(layers) {
  return JSON.stringify(
    layers.map(({ n, title, body }) => ({ layer: n, title, body: body.trim() })),
    null,
    2
  );
}

function toXml(layers) {
  const esc = (s) => String(s).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
  const parts = ['<?xml version="1.0" encoding="UTF-8"?>', "<prompt>"];
  for (const { n, title, body } of layers) {
    parts.push(`  <layer n="${n}" title="${esc(title)}"><![CDATA[\n${body}\n]]></layer>`);
  }
  parts.push("</prompt>");
  return parts.join("\n");
}

function main() {
  const { input, format, output } = parseArgs(process.argv);
  if (!input || !output) {
    console.error("usage: convert_format.js --input <yaml> --format md|json|xml|yaml --output <file>");
    process.exit(2);
  }
  const text = fs.readFileSync(input, "utf8");
  let out;
  if (format === "yaml") {
    out = text;
  } else {
    const layers = parseLayers(text);
    if (layers.length === 0) {
      console.error("[ERROR] Layer マーカー (# Layer N:) が見つかりません。正規形YAMLか確認してください。");
      process.exit(1);
    }
    if (format === "md") out = toMd(layers);
    else if (format === "json") out = toJson(layers);
    else if (format === "xml") out = toXml(layers);
    else { console.error(`unknown format: ${format}`); process.exit(2); }
  }
  fs.mkdirSync(path.dirname(output), { recursive: true });
  fs.writeFileSync(output, out);
  console.log(`converted → ${output} (${format})`);
}

main();

File: plugins/prompt-creator/skills/run-prompt-creator-7layer/scripts/convert-format.py
#!/usr/bin/env python3
# /// script
# name: convert-format
# purpose: 7層正規形 YAML を構造保存のまま md/json/xml/yaml へ変換する
# inputs:
#   - argv: --input <yaml> --format md|json|xml|yaml --output <file>
#   - file: --input の正規形 YAML
# outputs:
#   - file: --output に変換結果を書き出す
#   - stdout: converted → <output> (<format>)
#   - exit: 0=OK / 1=Layerマーカー不在 / 2=引数・未知format
# contexts: [C]
# network: false
# write-scope: output-arg-only
# dependencies: []
# ///
# convert-format.py — 7層正規形YAML → md/json/xml/yaml 変換 (Python 標準のみ)
# 設計方針: 日本語7層・ゴールシーク構造を「構造保存」で変換する。
#   YAML 本文（key: value / リスト / - [ ] チェックリスト / 達成ゴール 等）を捨てず保持し、
#   Layer 見出しのみ提示形式に合わせて書き換える。これにより可逆で、ゴールシーク要素を落とさない。
# マーカーは scaffold-prompt.py / merge-layers.py が出力する「# Layer N: <title>」に一致させる。
"""convert_format.js の python 移植。元の変換ロジック・終了コードを維持する。"""
import argparse
import json
import os
import re
import sys


def parse_args():
    parser = argparse.ArgumentParser(add_help=True)
    parser.add_argument("--input")
    parser.add_argument("--format", default="md")
    parser.add_argument("--output")
    # A4-10: parse_known_args の黙殺を廃止 (failfast)。未知引数は argparse が exit 2。
    return parser.parse_args()


# 「# Layer N: <title>」マーカーで本文を 7 層に分割し、本文を verbatim 保持する。
def parse_layers(text):
    layers = []
    for n in range(1, 8):
        pattern = (
            rf"#+\s*Layer\s*{n}\s*[:：]\s*([^\n]*)\n([\s\S]*?)"
            rf"(?=#+\s*Layer\s*{n + 1}\s*[:：]|$)"
        )
        m = re.search(pattern, text)
        if m:
            title = m.group(1).strip()
            body = re.sub(r"\s+$", "", m.group(2))
            layers.append({"n": n, "title": title, "body": body})
    return layers


def to_md(layers):
    out = ["# 7層構造プロンプト", ""]
    for layer in layers:
        out.append(f"## Layer {layer['n']}: {layer['title']}")
        out.append("")
        if layer["body"].strip():
            out.append(layer["body"].rstrip())
            out.append("")
    return "\n".join(out)


def to_json(layers):
    return json.dumps(
        [{"layer": x["n"], "title": x["title"], "body": x["body"].strip()} for x in layers],
        ensure_ascii=False,
        indent=2,
    )


def to_xml(layers):
    def esc(s):
        return str(s).replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")

    parts = ['<?xml version="1.0" encoding="UTF-8"?>', "<prompt>"]
    for layer in layers:
        parts.append(
            f'  <layer n="{layer["n"]}" title="{esc(layer["title"])}">'
            f'<![CDATA[\n{layer["body"]}\n]]></layer>'
        )
    parts.append("</prompt>")
    return "\n".join(parts)


def main():
    args = parse_args()
    input_path = args.input
    fmt = args.format
    output = args.output
    if not input_path or not output:
        sys.stderr.write(
            "usage: convert-format.py --input <yaml> --format md|json|xml|yaml --output <file>\n"
        )
        sys.exit(2)
    with open(input_path, "r", encoding="utf-8") as f:
        text = f.read()
    if fmt == "yaml":
        out = text
    else:
        layers = parse_layers(text)
        if len(layers) == 0:
            sys.stderr.write(
                "[ERROR] Layer マーカー (# Layer N:) が見つかりません。正規形YAMLか確認してください。\n"
            )
            sys.exit(1)
        if fmt == "md":
            out = to_md(layers)
        elif fmt == "json":
            out = to_json(layers)
        elif fmt == "xml":
            out = to_xml(layers)
        else:
            sys.stderr.write(f"unknown format: {fmt}\n")
            sys.exit(2)
    out_dir = os.path.dirname(output)
    if out_dir:
        os.makedirs(out_dir, exist_ok=True)
    with open(output, "w", encoding="utf-8") as f:
        f.write(out)
    print(f"converted → {output} ({fmt})")


if __name__ == "__main__":
    main()
